numcosmo/math/ncm_stats_dist_nd_kde_studentt.c

```diff
@@ -0,0 +1,1266 @@
+  
+  _ncm_stats_dist_nd_kde_studentt_prepare_IM (self);
+	ret = ncm_matrix_cholesky_decomp (self->IM, 'U');	
+
+	if (ret != 0)
+	{
+    _ncm_stats_dist_nd_kde_studentt_prepare_IM (self);
+
+    g_array_set_size (self->ipiv, self->n);
+    
+    ret = ncm_lapack_dsytrf ('U', self->n, ncm_matrix_data (self->IM), ncm_matrix_tda (self->IM), &g_array_index (self->ipiv, gint, 0), self->lapack_ws);
+    if (ret != 0)
+    {
+      printf ("ERR DEC2 h % 22.15g err % 22.15e %d\n", h, 1.0e20, ret);
+      return 1.0e20;
+    }
+
+    ncm_vector_memcpy (self->zeta, self->weights);
+    ret = ncm_lapack_dsytrs ('U', self->n, 1, ncm_matrix_data (self->IM), ncm_matrix_tda (self->IM), &g_array_index (self->ipiv, gint, 0), ncm_vector_data (self->zeta), ncm_vector_len (self->zeta));
+    if (ret != 0)
+      g_error ("_ncm_stats_dist_nd_kde_studentt_LOOCV_err2[ncm_lapack_dsytrs]: %d.", ret);
+    
+    ret = ncm_lapack_dsytri ('U', self->n, ncm_matrix_data (self->IM), ncm_matrix_tda (self->IM), &g_array_index (self->ipiv, gint, 0), self->lapack_ws);
+    if (ret != 0)
+      g_error ("_ncm_stats_dist_nd_kde_studentt_LOOCV_err2[ncm_lapack_dsytrs]: %d.", ret);
+	}
+  else
+  {
+    ncm_vector_memcpy (self->zeta, self->weights);
+    ret = ncm_matrix_cholesky_solve2 (self->IM, self->zeta, 'U');
+    if (ret != 0)
+      g_error ("_ncm_stats_dist_nd_kde_studentt_LOOCV_err2[ncm_matrix_cholesky_solve2]: %d.", ret);
+
+    ret = ncm_matrix_cholesky_inverse (self->IM, 'U');
+    if (ret != 0)
+      g_error ("_ncm_stats_dist_nd_kde_studentt_LOOCV_err2[ncm_matrix_cholesky_inverse]: %d.", ret);
+  }
+  
+	for (i = 0; i < self->n; i++)
+	{
+		err += gsl_pow_2 (ncm_vector_get (self->zeta, i) / (ncm_matrix_get (self->IM, i, i)));
+	}
+
+  return log (err);
+} 
+
+gdouble 
+_ncm_stats_dist_nd_kde_studentt_LOOCV_err2_f (guint n, const gdouble *x, gdouble *grad, gpointer f_data)
+{
+  const gdouble fx = _ncm_stats_dist_nd_kde_studentt_LOOCV_err2 (x[0], f_data);
+  /*printf ("# % 22.15g % 22.15g\n", x[0], fx);*/
+  return fx;
+}
+
+gdouble 
+_ncm_stats_dist_nd_kde_studentt_LOOCV_err2_full_f (guint n, const gdouble *x, gdouble *grad, gpointer f_data)
+{
+  NcmStatsDistNdKDEStudenttPrivate * const self = f_data;
+  gdouble err       = 0.0;
+	guint i, j, k;
+	gint ret;
+  gdouble tr_log_cov = 0.0;
+
+  k = 0;
+  for (i = 0; i < self->d; i++)
+  {
+    for (j = i; j < self->d; j++)
+    {
+      ncm_matrix_set (self->log_cov, i, j, x[k]);
+      if (i == j)
+        tr_log_cov += x[k];
+      k++;
+    }
+  }
+  ncm_matrix_sym_exp_cholesky (self->log_cov, 'U', self->cov_decomp);
+  _ncm_stats_dist_nd_kde_studentt_prepare_cov (self);
+  
+  self->href      = 1.0;
+  self->href2     = self->href * self->href;
+  self->us_lnnorm = log(gsl_sf_gamma((self->st_freedom + self->d) / 2)) - log(gsl_sf_gamma(self->st_freedom / 2)) - (self->d / 2) * (log(M_PI) + log(self->st_freedom)) -(1/2) * ncm_matrix_cholesky_lndet (self->cov_decomp);
```

Comment:
Since this computation is repeated here, it may be a good idea to create an internal (static) function to do this.